compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2426,7 +2426,14 @@ fn add_upstream_native_libraries(
                 NativeLibKind::Dylib { as_needed } => {
                     cmd.link_dylib(name, verbatim, as_needed.unwrap_or(true))
                 }
-                NativeLibKind::Unspecified => cmd.link_dylib(name, verbatim, true),
+                NativeLibKind::Unspecified => {
+                    cmd.link_dylib(
+                        name,
+                        verbatim,
+                        // FIXME: lld doesn't include some needed shared libraries without this, but includes some unneeded ones with this
+                        sess.opts.cg.link_args.iter().all(|arg| arg != "-fuse-ld=lld"),
```

Comment:
If no symbols from a dylib are used by the current dso, like I suspect is the case for you, `--as-needed` will not link the dylib. According to https://rust-lang.github.io/rfcs/2951-native-link-modifiers.html we may actually want to extend this behavior to all platforms by default. It provides the `-as-needed` linkage modifier to explicitly use `--no-as-needed`. It is not yet stabilized though.

Reply:
Why would the symbols not be used? What behavior should be extended to all platforms by default? Using `--as-needed`? I didn't know my change was platform specific somehow. That RFC definitely went over my head lol.

Reply:
I think only symbols from `mkl_core` are actually used by the binary. I think `mkl_core` itself depends on symbols provided by `mkl_intel_ilp64` or `mkl_intel_lp64` and `mkl_intel_thread` or `mkl_sequential` without declaring a dependency on any of them. `--as-needed` doesn't add a dependency on the latter libraries as they aren't used by the binary. It just so happens that during dynamic linking at program startup symbols provided by the latter libraries are actually used by the `mkl_core` version used at runtime. The `mkl_core` version used at runtime may be different from the version used while linking and as such have different symbol dependencies. They may even be provided by completely different dynamic libraries. As for why it works at all with ld.bfd (the default linker on most linux systems) I have no clue.

> What behavior should be extended to all platforms by default? Using --as-needed?

Yes

> That RFC definitely went over my head lol.

The RFC allows you to specify how to link against libraries. In you case you would need something like:

```rust
#[link(name = "mkl_intel_ilp64", modifiers = "-as-needed")]
extern "C" {}
#[link(name = "mkl_intel_thread", modifiers = "-as-needed")]
extern "C" {}
```

or equivalent in the build script:

```rust
println!("cargo:rustc-link-lib=dylib:-as-needed=mkl_intel_ilp64");
println!("cargo:rustc-link-lib=dylib:-as-needed=mkl_intel_thread");
```

Unfortunately this hasn't been stabilized yet.